README.md
Topi<img width="11%" align="right" src="https://github.com/caltechlibrary/topi/raw/main/.graphics/topi-icon.png">
======

Topi (_"**T**IND **O**bject **P**ython **I**nterface"_) is a Python package for getting basic data from a TIND server.

[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg?style=flat-square)](https://choosealicense.com/licenses/bsd-3-clause)
[![Latest release](https://img.shields.io/github/v/release/caltechlibrary/topi.svg?style=flat-square&color=b44e88)](https://github.com/caltechlibrary/topi/releases)
[![DOI](https://img.shields.io/badge/dynamic/json.svg?label=DOI&style=flat-square&colorA=gray&colorB=navy&query=$.metadata.doi&uri=https://data.caltech.edu/api/record/2105)](https://data.caltech.edu/records/2105)
[![Python](https://img.shields.io/badge/Python-3.6+-brightgreen.svg?style=flat-square)](http://shields.io)
[![PyPI](https://img.shields.io/pypi/v/topi.svg?style=flat-square&color=orange)](https://pypi.org/project/topi/)


Table of contents
-----------------

* [Introduction](#introduction)
* [Installation](#installation)
* [Usage](#usage)
* [Known issues and limitations](#known-issues-and-limitations)
* [Getting help](#getting-help)
* [Contributing](#contributing)
* [License](#license)
* [Acknowledgments](#authors-and-acknowledgments)


Introduction
------------

[TIND](https://tind.io) is an [integrated library system](https://en.wikipedia.org/wiki/Integrated_library_system). The Caltech Library uses a hosted solution by TIND for its [library catalog](https://caltech.tind.io).  Recent versions of TIND provide a REST API for getting a subset of information using network calls.  To make writing interfaces and automation scripts in Python easier, the Caltech Library [Digital Library Development team](https://www.library.caltech.edu/staff?&field_directory_department%5B0%5D=754) developed Topi (_"**T**IND **O**bject **P**ython **I**nterface"_), a Python package that provides an object-oriented interface to data in a TIND catalog.

Topi is not a complete API for interacting with TIND instances.  At this time, it provides an interface for retrieving only  two kinds of objects: bibliographic records, and items/holdings associated with those records.

A "topi" is also a [species of antelope](https://en.wikipedia.org/wiki/Topi) found in Africa.  The topi is currently classified as [_vulnerable_](https://en.wikipedia.org/wiki/Vulnerable_species) by the [International Union for Conservation of Nature](https://www.iucn.org) (IUCN) due to [threats](https://www.iucnredlist.org/species/6235/50185422) that include human development, hunting and droughts.


Installation
------------

The instructions below assume you have a Python interpreter installed on your computer; if that's not the case, please first [install Python version 3](INSTALL-Python3.md) and familiarize yourself with running Python programs on your system.

On **Linux**, **macOS**, and **Windows** operating systems, you should be able to install `topi` with [`pip`](https://pip.pypa.io/en/stable/installing/).  To install `topi` from the [Python package repository (PyPI)](https://pypi.org), run the following command:
```
python3 -m pip install topi
```

As an alternative to getting it from [PyPI](https://pypi.org), you can use `pip` to install `topi` directly from GitHub, like this:
```sh
python3 -m pip install git+https://github.com/caltechlibrary/topi.git
```

Usage
-----

Topi is a application programming interface (API) library; it does not offer a command-line interface.  There are three main option classes in Topi: `Tind`, `TindRecord`, and `TindItem`.  The rest of this section describes these classes and how to use them.

### Object classes

#### `Tind`

An object of the `Tind` class serves as the main point of interaction with a TIND server.  The constructor for `Tind` takes only one argument: the base network URL for the server.  Using it is very simple:

```python
from topi import Tind

tind = Tind('https://caltech.tind.io')
```

An instance of the `Tind` class offers just two methods: `record`, to create `TindRecord` objects, and `item`, to create `TindItem` objects.  These object classes are described below.


#### `TindRecord`

This object class represents a bibliographic record in a TIND database.  The fields of the record are derived from the MARC representation of the bibliographic record in TIND.  The following are the fields in a record object in Topi:

| Field name      | Type   | Description                                             |
|-----------------|--------|---------------------------------------------------------|
| `tind_id`       | string | The TIND record identifier                              |
| `tind_url`      | string | The URL to the online record page in TIND               |
| `title`         | string | The title (derived from MARC data field 245)            |
| `subtitle`      | string | The subtitle (derived from MARC data field 245)         |
| `author`        | string | The author(s) (derived from MARC data field 245 or 100) |
| `edition`       | string | The edition (derived from MARC data field 250)          |
| `publisher`     | string | The publisher's name (extracted from MARC control field 260) |
| `year`          | string | The year (extracted from MARC control field 008)        |
| `isbn_issn`     | list   | ISBN or ISSN numbers (from MARC data field 020)         |
| `description`   | string | A description, concatenated from MARC data field 300    |
| `bib_note`      | string | The value of MARC data field 504, subfield "a"          |
| `thumbnail_url` | string | The URL of the cover image in TIND (if any)             |
| `items`         | list   | A list of `TindItem` objects                            |

A `TindRecord` object can be obtained using the factory method `record(...)` on the `Tind` interface object.  This method takes one of two mutually-exclusive keyword arguments: either a TIND record identifier, or a MARC XML string obtained from a TIND server for a TIND bibliographic record.  Here is an example:

```python
from topi import Tind

tind = Tind('https://caltech.tind.io')
rec  = tind.record(tind_id = 680311)
```

Note the use of the keyword argument.  Below is an example of how to create a record from an existing MARC XML file obtained from a TIND server some other way &ndash; let's assume it is stored a file named `downloaded-marc.xml`:

```python
from topi import Tind

with open('downloaded_marc.xml', 'r') as xf:
    xml_string = xf.read()

tind = Tind('https://caltech.tind.io')
rec  = tind.record(marc_xml = xml_string)
```

The `thumbnail_url` field is lazily evaluated: its value is only obtained from the TIND server the first time the field is accessed by a calling program.  This is more efficient for situations where the thumbnail is never needed by an application, but it does mean that there is a delay the first time the field is accessed.


#### `TindItem`
    
Conceptually, in TIND an "item" is a specific copy of a work; this copy has a barcode and other item-specific information such as a location.  Each item is associated with a TIND record (represented by a `TindRecord` in Topi, described above).  The following are the fields in an item object in Topi:

| Field name    | Type         | Description                                         |
|---------------|--------------|-----------------------------------------------------|
| `parent`      | `TindRecord` | The parent record for this item                     |
| `barcode`     | string       | The item's barcode                                  |
| `type`        | string       | The type of item this is (e.g., "book")             |
| `volume`      | string       | The volume                                          |
| `call_number` | string       | The call number                                     |
| `description` | string       | A description of the specific item (e.g., "copy 1") |
| `library`     | string       | The library where the item is located               |
| `location`    | string       | The location of the item in the library             |
| `status`      | string       | Status of the item listed in TIND                   |

With Topi, a `TindItem` object can be obtained using the factory method `item(...)` on the `Tind` interface object.  This method takes a single argument: a barcode.  Here is an example:

```python
from topi import Tind

tind = Tind('https://caltech.tind.io')
item = tind.item(35047018228114)
```

Item records have parent pointers to the corresponding bibliographic record, in the form of a `TindRecord`.  Thus, given an item object, it's possible to look up the rest of the bibliographic record simply by dereferencing the `.parent` field:

```python
from topi import Tind

tind = Tind('https://caltech.tind.io')
item = tind.item(35047018228114)
print(item.parent.title)
```

Calling the `item` method on `Tind` will return an empty `TindItem` object.


### Additional notes

Topi fills out the `thumbnail_url` field of a `TindRecord` object by using TIND's API for the purpose.  This only retrieves what a given TIND database contains for the cover image of a work.  Other sources such as the [Open Library Covers API](https://openlibrary.org/dev/docs/api/covers) may have cover images that a TIND database lacks, but it is outside the scope of Topi to provide an interface for looking outside the TIND database.


Known issues and limitations
------------------------------

Currently, the coverage of the fields in `TindRecord` is limited.  Not all fields of a MARC XML record are mapped to fields in `TindRecord` at this time.  (Code contributions are welcome!)


Getting help
------------

If you find an issue, please submit it in [the GitHub issue tracker](https://github.com/caltechlibrary/topi/issues) for this repository.


Contributing
------------

We would be happy to receive your help and participation with enhancing Topi!  Please visit the [guidelines for contributing](CONTRIBUTING.md) for some tips on getting started.


License
-------

Software produced by the Caltech Library is Copyright (C) 2021, Caltech.  This software is freely distributed under a BSD/MIT type license.  Please see the [LICENSE](LICENSE) file for more information.


Acknowledgments
---------------

This work was funded by the California Institute of Technology Library.

The [vector artwork](https://thenounproject.com/term/antelope/931009/) of an antelope, used as the icon for this repository, was created by [parkjisun](https://thenounproject.com/naripuru/) for the Noun Project.  It is licensed under the Creative Commons [CC-BY 3.0](https://creativecommons.org/licenses/by/3.0/) license.

Topi makes use of numerous open-source packages, without which Topi could not have been developed.  I want to acknowledge this debt.  In alphabetical order, the packages are:

* [commonpy](https://github.com/caltechlibrary/commonpy) &ndash; a collection of commonly-useful Python functions
* [cssselect](https://pypi.org/project/cssselect/) &ndash; `lxml` add-on to parse CSS3 selectors 
* [ipdb](https://github.com/gotcha/ipdb) &ndash; the IPython debugger
* [lxml](https://lxml.de) &ndash; an XML parsing library for Python
* [setuptools](https://github.com/pypa/setuptools) &ndash; library for `setup.py`
* [sidetrack](https://github.com/caltechlibrary/sidetrack) &ndash; simple debug logging/tracing package


<div align="center">
  <br>
  <a href="https://www.caltech.edu">
    <img width="100" height="100" src="https://raw.githubusercontent.com/caltechlibrary/topi/main/.graphics/caltech-round.png">
  </a>
</div>


CHANGES.md
# Change log for Topi

## Version 1.1.0

* Add `publisher` field to TIND record object.
* Fix serious bugs in `tind.py`.
* Update test cases to account for changes in entries in Caltech's TIND catalog.


## Version 1.0.2

* Fix bugs in `tind_utils.py`.
* Update required version of [Sidetrack](https://github.com/caltechlibrary/sidetrack) to 2.0.


## Version 1.0.1

* Change approach for setting `tind_url` field on `TindRecord`. (Somehow the test cases previously didn't catch that it wasn't being set at all.)


## Version 1.0.0

* Removed build files accidentally committed during last release.
* Added more info to the [`README` file](README.md).
* Some internal code improvements.


## Version 0.0.2

Renamed the project to Topi (TIND Object Python Interface) to avoid a name clash on PyPI for the name "Potion".


## Version 0.0.1

First release.


topi/tind.py
'''
tind.py: code for interacting with a TIND server

This downloads data in the MARC XML format instead of other formats, because
it contains data not present in the other output formats.  In addition,
TIND doesn't seem to do any additional cleanup on the data in the other
formats, so there's no real advantage. E.g., I find title values in the other
formats ending with "/", some fields have trailing commas in their values, etc.

Authors
-------

Michael Hucka <mhucka@caltech.edu> -- Caltech Library

Copyright
---------

Copyright (c) 2021 by the California Institute of Technology.  This code
is open-source software released under a 3-clause BSD license.  Please see the
file "LICENSE" for more information.
'''

from   collections import namedtuple
from   commonpy.network_utils import net
from   commonpy.exceptions import NoContent, ServiceFailure, RateLimitExceeded
import json
from   json import JSONDecodeError
from   lxml import etree

if __debug__:
    from sidetrack import log

from .exceptions import *
from .item import TindItem
from .tind_utils import result_from_api
from .record import TindRecord


# Constants.
# .............................................................................

# Field names for finding different elements in a MARC XML record.
ELEM_RECORD       = '{http://www.loc.gov/MARC21/slim}record'
ELEM_CONTROLFIELD = '{http://www.loc.gov/MARC21/slim}controlfield'
ELEM_DATAFIELD    = '{http://www.loc.gov/MARC21/slim}datafield'
ELEM_SUBFIELD     = '{http://www.loc.gov/MARC21/slim}subfield'

# URL templates for retrieving a record from a TIND server.
# The first placeholder is for the host URL; the second is for an identifier.
# Use Python .format() to substitute the relevant values into the string.
_MARCXML_FOR_BARCODE = '{}/search?p=barcode%3A+{}&of=xm'
_MARCXML_FOR_TIND_ID = '{}/search?recid={}&of=xm'

# URL template for item data from a TIND server.
# The first placeholder is for the host URL; the second is for a TIND record id.
# Use Python .format() to substitute the relevant values into the string.
_ITEMS_FOR_TIND_ID = '{}/nanna/bibcirc/{}/details'


# Class definitions.
# .............................................................................

class Tind():
    '''Interface to a TIND.io server.'''

    def __init__(self, server_url):
        '''Create an interface to the TIND server at "server_url".'''
        self.server_url = server_url


    def record(self, tind_id = None, marc_xml = None):
        '''Create a TindRecord object given either a TIND id or MARC XML.

        Keyword arguments "tind_id" and "marc_xml" are mutually exclusive.

        If "tind_id" is given, this will search the TIND server using the id,
        download the MARC XML returned by TIND, create a TindRecord object
        based on the data, then separately, also use TIND's TindItem record
        API to find all the items/holdings for the record, create
        corresponding TindItem objects for each, and add them to the "items"
        list within the TindRecord object. If the TIND server does not return
        a result for the id, this method raises a NotFound exception.

        If "marc_xml" is given, the XML must have been obtained using the
        MARC XML export feature of TIND.  This method will skip the
        preliminary TIND search and instead, create the record using the
        given XML string, then perform the same steps as described above for
        the case of tind_id.

        If neither "tind_id" nor "marc_xml" is given, this method returns an
        empty TindRecord object.
        '''
        if tind_id and marc_xml:
            raise ValueError(f'"tind_id" and "marc_xml" are mutually exclusive.')

        if tind_id:
            tind_id = str(tind_id)
            if not tind_id.isdigit():
                raise ValueError(f'Invalid argument: {tind_id} is not a number.')
            record = self._record_from_server(_MARCXML_FOR_TIND_ID, tind_id)
        elif marc_xml:
            if not marc_xml.startswith(b'<?xml'):
                raise ValueError(f'marc_xml argument does not appear to be XML.')
            record = self._record_from_xml(marc_xml)
        else:
            return TindRecord(server_url = self.server_url)

        if record:
            record.items = self._items_for_tind_id(tind_id or record.tind_id)
            for item in record.items:
                item.parent = record
            return record
        else:
            arg = tind_id if tind_id else 'given XML data'
            raise NotFound(f'No record found for {arg} in {self.server_url}')


    def item(self, barcode = None):
        '''Create a TindItem object given a barcode value.

        This will contact the TIND server and perform a search using the
        barcode value, then using the data returned, create a TindRecord
        object with an TindItem object within it, and finally return the
        TindItem object.  (TindItem objects contain a property referencing
        their parent TindRecord objects, so callers can get the TindRecord
        given the TindItem object.)

        If the TIND server does not return a result for the barcode, this
        method raises a NotFound exception.

        If no barcode is given, this returns an empty TindItem object.
        '''
        if not barcode:
            return TindItem()
        barcode = str(barcode)
        if not barcode.isdigit():
            raise ValueError(f'Invalid argument: {barcode} is not a number.')
        record = self._record_from_server(_MARCXML_FOR_BARCODE, barcode)
        if record:
            record.items = self._items_for_tind_id(record.tind_id)
            for item in record.items:
                item.parent = record
            for item in record.items:
                if item.barcode == barcode:
                    return item
            raise DataMismatchError('Unable to match item to record from TIND.')
        else:
            raise NotFound(f'No record found for {barcode} in {self.server_url}')


    def _record_from_server(self, url_template, id):
        '''Create a TindRecord by contacting "url_template" with the "id".'''
        def response_handler(resp):
            if not resp or not resp.content:
                if __debug__: log(f'got no response for {endpoint}')
                return
            record = self._record_from_xml(resp.content)
            return record

        endpoint = url_template.format(self.server_url, id)
        return result_from_api(endpoint, response_handler)


    def _record_from_xml(self, xml):
        '''Initialize this record given MARC XML as a string.'''

        record = TindRecord(server_url = self.server_url)
        # Save the XML internally in case it's useful.
        record._xml = xml

        # Parse the XML.
        if __debug__: log(f'parsing MARC XML {len(xml)} chars long')
        try:
            parser = etree.XMLParser(recover = True)
            tree = etree.fromstring(xml, parser = parser)
        except Exception as ex:
            raise ValueError(f'Bad XML')
        if len(tree) == 0:             # Blank record.
            if __debug__: log(f'blank record -- no values parsed')
            return record
        elements = tree.find(ELEM_RECORD)
        subtitle = None
        for element in elements.findall(ELEM_CONTROLFIELD):
            if element.attrib['tag'] == '001':
                record.tind_id = element.text.strip()
            elif element.attrib['tag'] == '008':
                record.year = element.text[7:11].strip()
                if not record.year.isdigit():
                    record.year = ''

        main_author = None
        for element in elements.findall(ELEM_DATAFIELD):
            if element.attrib['tag'] == '250':
                record.edition = element.find(ELEM_SUBFIELD).text.strip()
            elif element.attrib['tag'] == '050':
                record.call_no = ''
                for subfield in element.findall(ELEM_SUBFIELD):
                    record.call_no += subfield.text.strip() + ' '
            elif element.attrib['tag'] == '100':
                for subfield in element.findall(ELEM_SUBFIELD):
                    if subfield.attrib['code'] == 'a':
                        main_author = subfield.text.strip()
            elif element.attrib['tag'] == '245':
                for subfield in element.findall(ELEM_SUBFIELD):
                    if subfield.attrib['code'] == 'a':
                        text = subfield.text.strip()
                        # The title sometimes contains the author names too.
                        record.title, record.author = parsed_title_and_author(text)
                    elif subfield.attrib['code'] == 'b':
                        record.subtitle = subfield.text.strip()
                    elif subfield.attrib['code'] == 'c':
                        record.author = subfield.text.strip()
            elif element.attrib['tag'] == '020':
                for subfield in element.findall(ELEM_SUBFIELD):
                    # Value is sometimes of the form "1429224045 (hbk.)"
                    value = subfield.text.split()[0]
                    if value.isdigit():
                        record.isbn_issn.append(value)
            elif element.attrib['tag'] == '300':
                parts = [sub.text.strip() for sub in element.findall(ELEM_SUBFIELD)]
                record.description = ' '.join(parts)
            elif element.attrib['tag'] == '504':
                for subfield in element.findall(ELEM_SUBFIELD):
                    if subfield.attrib['code'] == 'a':
                        record.note = subfield.text.strip()
            elif element.attrib['tag'] == '260':
                for subfield in element.findall(ELEM_SUBFIELD):
                    if subfield.attrib['code'] == 'b':
                        record.publisher = subfield.text.strip()

        # We get author from 245 because in our entries, it's frequently part
        # of the title statement. If it's not, but we got an author from 100
        # use that.  100 only lists first author, but it's better than nothing.
        if record.author:
            if record.author.startswith('by'):
                record.author = record.author[2:].strip()
            elif record.author.startswith('edited by'):
                record.author = record.author[10:].strip()
        elif main_author:
            record.author = main_author

        # Caltech's TIND database contains some things that are not reading
        # materials per se. The following is an attempt to weed those out.
        if sum([not record.author, not record.year, not record.title]) > 1:
            for field in ['title', 'author', 'year', 'call_no', 'edition']:
                setattr(record, field, None)
            return record

        # Some cleanup work is better left until after we obtain all values.
        record.author      = cleaned(record.author)
        record.title       = cleaned(record.title)
        record.edition     = cleaned(record.edition)
        record.subtitle    = cleaned(record.subtitle)
        record.description = cleaned(record.description)
        record.publisher   = cleaned(record.publisher)

        return record


    def _items_for_tind_id(self, id):
        '''Return a list of TindItem objects for the TIND record "id".'''
        def response_handler(resp):
            results = []
            if not resp or not resp.text:
                return []
            try:
                data = json.loads(resp.text)
            except JSONDecodeError as ex:
                raise TindError(f'Malformed result from {self.server_url}: str(ex)')
            except TypeError as ex:
                raise DataMismatchError(f'Unexpected data returned by {self.server_url}.')

            if 'items' not in data:
                if __debug__: log(f'results from server missing "items" key')
                raise TindError(f'Unexpected result from {self.server_url}')
            for item in data['items']:
                results.append(TindItem(barcode     = item.get('barcode', ''),
                                        type        = item.get('item_type', ''),
                                        volume      = item.get('item_volume', ''),
                                        call_number = item.get('call_number', ''),
                                        description = item.get('description', ''),
                                        library     = item.get('library', '',),
                                        location    = item.get('location', ''),
                                        status      = item.get('status', '')))
            return results

        endpoint = _ITEMS_FOR_TIND_ID.format(self.server_url, id)
        return result_from_api(endpoint, response_handler)


# Miscellaneous helpers.
# .............................................................................

def cleaned(text):
    '''Mildly clean up the given text string.'''
    if not text:
        return text
    text = text.rstrip('./,')
    return text.strip()


def parsed_title_and_author(text):
    '''Extract a title and authors (if present) from the given text string.'''
    title = None
    author = None
    if text.find('/') > 0:
        start = text.find('/')
        title = text[:start].strip()
        author = text[start + 3:].strip()
    elif text.find('[by]') > 0:
        start = text.find('[by]')
        title = text[:start].strip()
        author = text[start + 5:].strip()
    elif text.rfind(', by') > 0:
        start = text.rfind(', by')
        title = text[:start].strip()
        author = text[start + 5:].strip()
    else:
        title = text
    if title.endswith(':'):
        title = title[:-1].strip()
    return title, author
